File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public void backTrack(int index,List<String> list,String digits,StringBuilder sb,
                         String arr[])
    {
        if(sb.length()==digits.length())
        {
            list.add(sb.toString());
            return ;
        }
        else
        {
            int x=digits.charAt(index)-'0';
            for(int i=0;i<arr[x].length();i++)
            {
                sb.append(arr[x].charAt(i));
                backTrack(index+1,list,digits,sb,arr);
                sb.deleteCharAt(sb.length()-1);
            }
        }
    }
    public List<String> letterCombinations(String digits) {
        List<String> list=new ArrayList<>();
        if(digits.length()==0)
            return list;
        String arr[]=new String[]{"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        backTrack(0,list,digits,new StringBuilder(),arr);
        return list;
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<std::string> list;
    string arr[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    void helper(int idx, string digits, string combination) {
        if (idx == digits.size()) {
            list.push_back(combination);
            return;
        }

        int x = digits[idx] - '0';
        for (int i = 0; i < arr[x].size(); i++) {
            combination.push_back(arr[x][i]);
            helper(idx + 1, digits, combination);
            combination.pop_back();
        }
    }

    vector<string> letterCombinations(string digits) {
        if(digits.size()==0)
            return list;
        helper(0,digits,"");
        return list;
    }
};